### ISSUE TITLE
Aggregate output of validation_end across all ddp processes

### ISSUE THREAD

**Issue by xiadingZ** [PyTorchLightning/pytorch-lightning#243]
In `validation_end` I want to gather all outputs from `validation_step`, then use another file to calculate scores. but when use multi-gpu ddp mode, I find it launches multiple process of `validation_end` to calculate scores. how to make it only call once?

**Comment by williamFalcon** [comment]
hi, the way ddp works, each process is walled from the other. I’m not sure you can transfer arbitrary tensors around but we can look into it (likely using the dist library). 

the validation_end will be called for every process to calculate all the scores.

I guess it would be helpful to know what you are trying to do to see how we modify the code.

**Comment by xiadingZ** [comment]
This is what I want to do:
  every process of validation_step computes scores of corresponding data from dataloader, then in validation_end, a main process gather all outputs from all validation_steps to form a score of all data, then use a gt_file to compute auc of all data, only once.

**Comment by xiadingZ** [comment]
It seems that I want to do test in val function. I find a workaround to do this only when `self.trainer.proc_rank ==0`, in validation_end. 
thanks

**Comment by williamFalcon** [comment]
so, let’s break this down a bit. 

Case 1 (current):
Each process calls validation_end. If you log anything or monitor early stopping it only uses the metrics from proc 0.  This is equivalent to using 1/nb_process in this calculation. 

Case 2 (i think this is the one we need to support):
Same as above, but use dist.all_reduce for each metric returned from validation_end to get the full value instead of the estimated 1/nb_process value. 

example:
Case 1:
you calculate accuracy. then the accuracy you log (only from proc 0) is an estimate of the full accuracy. 

Case 2:
you calculate accuracy, then the dist.all_reduce is not an estimate but the full accuracy. 

If your data are uniformly shuffled and your batches are big enough, case 1 and 2 are almost identical. if your batch is too small, the estimate will be a little bit off.

**Comment by neggert** [comment]
Agree that case 2 is important to support, and should maybe even be the default. That seems like what a new user would expect.

**Comment by williamFalcon** [comment]
Yeah, agreed. I'll turn this into a ticket.

**Comment by williamFalcon** [comment]
@neggert does it also make sense to do the same after training_step? I'm a bit concerned about the speed impact of adding these calls

**Comment by neggert** [comment]
I can't think of a good reason to do it after a training step. Any metrics measured batch-by-batch are going to be noisy anyway, so only taking numbers from one process shouldn't make much difference. It's possible there's some use case I haven't thought of, though.

**Comment by neggert** [comment]
I guess I'll also note that averaging across nodes isn't always going to be the right thing to do. Metric learning problems, for instance, actually need a full set of feature vectors to compute recall and NMI.

The cleanest thing IMO would be to collect `validation_step` outputs from all nodes using something along the lines of `all_gather`, then pass the whole thing to `validation_end` on each node. I'm not sure how to actually do this, though.

**Comment by williamFalcon** [comment]
FYI @adefazio.

@neggert Let's go for this behavior:
- reduce each metric on validation_step only.

**Comment by magic282** [comment]
Also having this issue. Temporarily disabling the distributedsampler so I can have a full validation set for each process.

**Comment by brucemuller** [comment]
Hello

@neggert @williamFalcon did an update happen for this? I'm also trying to aggregate validation_end between ddp GPU processes on a single node. I'm trying to use dist.all_gather or dist.all_gather_multigpu but really unsure how it is done with pytorch lightning. I also think new users would expect info on how to aggregate from multiple processes on GPUs

**Comment by williamFalcon** [comment]
moving discussion to #702
